hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/SecurityUtil.java

```diff
@@ -202,32 +201,89 @@ public static String getServerPrincipal(String principalConfig,
   public static String getServerPrincipal(String principalConfig,
       InetAddress addr) throws IOException {
     String[] components = getComponents(principalConfig);
-    if (components == null || components.length != 3
-        || !components[1].equals(HOSTNAME_PATTERN)) {
+    if (components == null || components.length != 3) {
       return principalConfig;
-    } else {
-      if (addr == null) {
-        throw new IOException("Can't replace " + HOSTNAME_PATTERN
-            + " pattern since client address is null");
-      }
-      return replacePattern(components, addr.getCanonicalHostName());
     }
+
+    String service = components[0];
+    String host = components[1];
+    String realm = components[2];
+
+    boolean hostIsPattern = host.equals(HOSTNAME_PATTERN);
+    boolean realmIsPattern = realm.equals(REALM_PATTERN);
+
+    if (!hostIsPattern && !realmIsPattern) {
+      return principalConfig;
+    }
+
+    if (addr == null) {
+      throw new IOException("Can't replace principal pattern since client address is null");
+    }
+
+    String canonicalHostName = addr.getCanonicalHostName();
+
+    if (hostIsPattern) {
+      host = getHostName(canonicalHostName);
+    }
+
+    if (realmIsPattern) {
+      realm = getCachedHostRealm(canonicalHostName);
+    }
+
+    return service + "/" + host + "@" + realm;
   }
-  
+
   private static String[] getComponents(String principalConfig) {
     if (principalConfig == null)
       return null;
     return principalConfig.split("[/@]");
   }
-  
-  private static String replacePattern(String[] components, String hostname)
+
+  private static String getHostName(String hostname)
       throws IOException {
     String fqdn = hostname;
     if (fqdn == null || fqdn.isEmpty() || fqdn.equals("0.0.0.0")) {
       fqdn = getLocalHostName(null);
     }
-    return components[0] + "/" +
-        StringUtils.toLowerCase(fqdn) + "@" + components[2];
+    return StringUtils.toLowerCase(fqdn);
+  }
+
+  private static String getCachedHostRealm(String hostname) throws IOException {
+    String realm = realmCache.get(hostname);
+    if (realm == null) {
+      realm = getHostRealm(hostname);
+      realmCache.put(hostname, realm);
+    }
+
+    return realm;
+  }
+
+  private static String getHostRealm(String hostname) throws IOException {
+    if (realmScript == null || realmScript.isEmpty()) {
+      RuntimeException e = new IllegalArgumentException(REALM_PATTERN +
+              " usage requires " + CommonConfigurationKeys.HADOOP_SECURITY_KERBEROS_REALM_SCRIPT_KEY + ". Check your" +
+              "configuration.");
+      LOG.error("server principal", e);
+      throw e;
+    }
+
+    try {
+      String[] env = {"CANONICAL_HOSTNAME=" + hostname};
+      Process proc = Runtime.getRuntime().exec(realmScript, env, null);
```

Comment:
Can you give a try to the hadoop common class ShellCommandExecutor ? Here again, mostly to stay close to other hadoop code.